### ISSUE TITLE
Store the Mimetype on storage backends supporting it

### ISSUE THREAD

**Issue by meidlinga** [tus/tusd#1016]
**Question**
I would like to include the uploaded files mime type in the upload meta data.
Then I would like to store this information along the uploaded file on Azure storage.
On a quick glance over the code, I could not find this feature.
Would this currently be possible in any way?

**Setup details**
Please provide following details, if applicable to your situation:
- Operating System: Linux
- Used tusd version: --
- Used tusd data storage: Azure
- Used tusd configuration: --
- Used tus client library: --

**Comment by Acconut** [comment]
A similar question also appeared for S3 in https://github.com/tus/tusd/issues/978. I think it would be worth adding this to storages, which provide some mechanism for saving the file name and file type in a dedicated manner. Does Azure have such a functionality?

**Comment by meidlinga** [comment]
Yes, Azure supports it by adding a specific header in the upload request. It is `x-ms-blob-content-type`, documented here: https://learn.microsoft.com/en-us/rest/api/storageservices/put-blob?tabs=azure-ad#request-headers-all-blob-types

**Comment by Acconut** [comment]
That's good to know. It is also very similar to the `Content-Type` header that is used for AWS S3 to set the file type for an object: https://docs.aws.amazon.com/AmazonS3/latest/API/API_CreateMultipartUpload.html#API_CreateMultipartUpload_RequestSyntax

It should not be too hard to set those two values when creating an upload. tusd currently inspects the `filetype` meta data entry and sets `Content-Disposition` accordingly when a user downloads a file via tusd. We can also read this value and pass it to Azure, S3 etc.

**Comment by Acconut** [comment]
It would be great if tusd could set Content-Type, Content-Encoding, and Content-Disposition for storages that support it.

**Comment by quality-leftovers** [comment]
Looking at the code (as of today) it seems s3store supports setting the blob `Content-Type` using either
-  `Content-Type` header when using TUS V2 draft protocol
- `filetype` metadata field using TUS V1 protocol

The s3store also attaches metadata to the blob while azurestore just ignores it.

So I did play around a bit with the azurestore. See: https://github.com/quality-leftovers/tusd/commit/4909c02a5096fc47d353b4499dd65f4cd86c7661

If there is interest in getting this into main I'll try to find some time to work on an actual PR.

**Comment by Acconut** [comment]
> Looking at the code (as of today) it seems s3store supports setting the blob `Content-Type`

Yes, you're correct. It has been implemented in https://github.com/tus/tusd/pull/1217 and I would more than happy if we can provide similar handling to other storages as well.

That being said, I'm unsure whether the meta data handling for S3 should serve as a basis for other storages as it has its problems. For example, meta data has to be sanitized or otherwise the storage provider might reject the object entirely due to invalid character. In addition, meta data from tus uploads are, by default, passed through to the S3 object, meaning that they can effectively inject custom meta data unless tusd hooks are employed to filter and restrict meta data.

Therefore, I would like to revisit this approach for meta data handling for S3 and other storages. I think in the future, tusd should not set object meta data by default to avoid the aforementioned problems. Instead, tusd should provide the tus upload meta data to the `pre-create`, which can then decide whay metadata to keep or not through a new setting in `FileInfoChanges.Storage["MetaData"]` field (https://pkg.go.dev/github.com/tus/tusd/v2/pkg/handler#FileInfoChanges). Whatever's set in there will be passed to the object storage as additional metadata. By default, this value is empty, meaning no object metadata. This is in contrast to `FileInfoChanges.MetaData`, which is the tus upload metadata, which is stored in the info objects.

In addition, I think tusd should also set Content-Type and Content-Disposition by default for all storages (if they support it), while also allowing `pre-create` hooks to overwrite it. Content-Type and Content-Disposition would be validated and sanitized by tusd, so users can be sure that they don't cause major harm.

What do you think about this approach? I've started exploring this in https://github.com/tus/tusd/pull/1167, but it's not yet finished.

**Comment by quality-leftovers** [comment]
Good points. I also think it would be nice if it was handled in the hooks.

What I had initially in mind was adding a seperate field in the `FileInfoChanges` named `StorageMetadata` or somehow putting the metadata into the existing `Storage` map. But since adding something in the object lead to lots of changes (GRPC hooks, tests) all over the place I just went with the S3 behavior.

Maybe putting all metadata into `FileInfoChanges.Storage["MetaData"]` would really be a good solution. We'd just need to define a nice encoding and have a `GetStorageMetadata` function somewhere that all stores can use that returns a `map[string]string`

**Comment by Acconut** [comment]
> We'd just need to define a nice encoding (maybe just use a json document?) and have a `GetStorageMetadata` function somewhere that all stores can use to get the actual metadata (`map[string]string`).

Not sure, I understand. What would the method be used for? My thinking is that `FileInfoChanges.Storage["MetaData"]` is used by the pre-create hook to specify metadata that should be added to the object that's saved in the underlying storage, if supported. There is no need to fetch this metadata later again in tusd. At least nothing I can think if right now. Please remember that this object metadata is different than the upload meta, which is controlled by `FileInfoChanges.MetaData`.

**Comment by quality-leftovers** [comment]
As far as I know for Azure you cannot create a "Block Blob" with metadata. Even if you create an empty blob unless you create it with empty content which should raise a `Microsoft.Storage.BlobCreated` I think the metadata will just be replaced when the blob is finally  committed.

So basically the "upload" flow for azure storage is:
- Create Info
- Put Content Blob Blocks (Chunks)
- Commit Content Blob Blocks => can specify metadata here
- Content Blob is committed and Azure Storage raises  `Microsoft.Storage.BlobCreated` event

Therefore azurestore implementation in tusd would need to persist metadata somewhere until content blob can be committed, unless someone has an idea how to handle block blobs "differently".
